lib/utils/signbjstx.js

```diff
@@ -0,0 +1,173 @@
+                                : hdnodeUtils.derivePubKeyHash(
+                                      nodes,
+                                      pathOrAddress[pathOrAddress.length - 2],
+                                      pathOrAddress[pathOrAddress.length - 1]
+                                );
+
+    if (scriptType === 'PAYTOADDRESS') {
+        return bitcoin.script.pubKeyHashOutput(pkh);
+    }
+    if (scriptType === 'PAYTOSCRIPTHASH') {
+        return bitcoin.script.scriptHashOutput(pkh);
+    }
+    throw new Error('Unknown script type ' + scriptType);
+}
+
+function verifyBjsTx(
+    inputs: Array<types.InputInfo>,
+    outputs: Array<types.OutputInfo>,
+    nodes: Array<bitcoin.HDNode>,
+    resTx: bitcoin.Transaction,
+    network: bitcoin.Network
+) {
+    if (inputs.length !== resTx.ins.length) {
+        throw new Error('Signed transaction has wrong length.');
+    }
+    if (outputs.length !== resTx.outs.length) {
+        throw new Error('Signed transaction has wrong length.');
+    }
+    outputs.map((output, i) => {
+        if (output.value !== resTx.outs[i].value) {
+            throw new Error('Signed transaction has wrong output value.');
+        }
+
+        const addressOrPath_: ?(string | Array<number>) =
+            output.address == null ? (output.path == null ? null : output.path) : output.address;
+
+        if (addressOrPath_ == null) {
+            throw new Error('Both path and address cannot be null.');
+        }
+
+        const addressOrPath: (string | Array<number>) = addressOrPath_;
+        const scriptA: Buffer =
+            deriveOutputScript(addressOrPath, nodes, network);
+        const scriptB: Buffer = resTx.outs[i].script;
+        if (scriptA.compare(scriptB) !== 0) {
+            throw new Error('Scripts differ');
+        }
+    });
+}
+
+function getAddressScriptType(address: string, network: bitcoin.Network): string {
+    const decoded = bitcoin.address.fromBase58Check(address);
+
+    if (decoded.version === network.pubKeyHash) {
+        return 'PAYTOADDRESS';
+    }
+
+    if (decoded.version === network.scriptHash) {
+        return 'PAYTOSCRIPTHASH';
+    }
+
+    throw new Error('Unknown address type.');
+}
+
+export function signBjsTx(
+    info: types.TxInfo,
+    refTxs: Array<bitcoin.Transaction>,
+    nodes: Array<bitcoin.HDNode>,
+    coinName: string
+): Promise<bitcoin.Transaction> {
+    const network: bitcoin.Network = bitcoin.networks[coinName.toLowerCase()];
+    if (network == null) {
+        return Promise.reject(new Error('No network ' + coinName));
+    }
+
+    const trezorInputs: Array<types.TransactionInput> = info.inputs.map(i => input2trezor(i));
+    const trezorOutputs: Array<types.TransactionOutput> =
+        info.outputs.map(o => output2trezor(o, network));
+    const trezorRefTxs: Array<types.RefTransaction> = refTxs.map(tx => bjsTx2refTx(tx));
+
+    return this.signTx(
```

Comment:
I'll edit this a bit later, as I think we shouldn't touch `this` outside the class. It could actually simplify the code a bit.